determined-ai/determined
# feat: sharded checkpoints

## Description
Sharded checkpoints allow to upload data from all nodes (not only from the master), and download/restore_path selected files to each node.

To upload sharded data, a user must invoke `upload` or `store_path` with `shard=True` from all workers. Before creating the checkpoint, workers create a list of all resources and check for conflicts. A conflict happens when more than one rank wants to upload a given file. In such cases, we terminate with an error. The same happens for metadata. Each worker can provide different metadata to upload/store_path. Metadata will be merged, and if a conflict is detected (i.e., a key in the metadata is repeated), the process terminates with an error.
For each node, only the lowest rank uploads resources.

For download/restore_path, a user may provide a selector: `Callable[[str], bool]`, which decides which files will be downloaded. If no selector is provided, each node will get all the files. If a selector is provided for each rank, local ranks must coordinate and create a merged list of resources to download locally.





## Test Plan
To test manually:
Upload/Store_path: run an experiment on two or more nodes and invoke upload/store_path with shard=True. For upload, make sure that each node has a different file in the uploaded (src) dir. For store_path, save different files in the provided path. After upload/store_path is done, manually inspected registered checkpoint to see files from all nodes being recorded.

Download/Restore_path: provide a selector `Callable[[str], bool]` for each rank to download only a subset of files. Inspect that only selected files are downloaded.

Tests:
* added tests for download/restore_path with selector to run_storage_lifecycle_test invoked for all storages
* added tests for s3, simulating two nodes doing upload/download and store_path/restore_path with sharded data.



## Commentary (optional)




## Checklist
- [ x] Changes have been manually QA'd
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
